File: C++/2583_영역구하기.cpp
#include <iostream>
#include <cstring>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int main() {
	int M, N, K;
	cin >> M >> N >> K;
	
	bool **area = new bool*[M];
	for(int i = 0; i < M; i++){
		area[i] = new bool[N];
		memset(area[i],false,N);
	}
	
	for(int i = 0; i < K; i++){ // 직사각형 영역의 area 값 true로 바꿔주기
		int x1, y1, x2, y2;
		cin >> x1 >> y1 >> x2 >> y2;
		y1 = M-y1;
		y2 = M-y2; // 문제는 왼쪽 아래가 (0,0) 코드에서는 왼쪽 위가 (0,0) -> y좌표의 값 y = M 대칭시켜줘야 함
		
		for(int y = y2; y < y1; y++){
			for(int x = x1; x < x2; x++){
				area[y][x] = true;
			}
		}
	}
	
	int dir[4][2] = {{-1,0}, {1,0}, {0,-1}, {0,1}}; // 상하좌우
	
	vector<int> n_area; // 영역의 넓이를 담을 벡터
	for(int i = 0; i < M; i++){
		for(int j = 0; j < N; j++){
			if(area[i][j] == false){ // 사각형이 채워져 있지 않거나 방문하지 않은 점을 시작으로
				queue<pair<int, int> > q;
				q.push(make_pair(i,j));
				area[i][j] = true; // bfs 수행
				
				int cnt = 0; // 현재 영역의 넓이
				while(!q.empty()){
					int y = q.front().first;
					int x = q.front().second;
					q.pop();
					
					cnt++;
          
					for(int d = 0; d < 4; d++){
						int dy = y+dir[d][0], dx = x+dir[d][1];
						if(dy >= 0 && dy < M && dx >= 0 && dx < N && area[dy][dx] == false){ // 영역 내에 존재해야 하고 방문한 적 없어야 함
							q.push(make_pair(dy,dx));
							area[dy][dx] = true;
						}
					}
					
				}
				
				n_area.push_back(cnt); // 현재 영역의 넓이 push
			}
		}
	}
	
	sort(n_area.begin(),n_area.end()); // 영역의 넓이를 오름 차순으로 정렬
	cout << n_area.size() << '\n'; // 벡터의 크기가 영역의 개수가 됨
	for(int i = 0; i < n_area.size(); i++){
		cout << n_area[i] << " "; // 정렬된 영역의 넓이 출력
	}
	
	return 0;
}

File: Python/2583_영역구하기.py
import sys
import collections

if __name__ == '__main__':
	M, N, K = list(map(int,sys.stdin.readline().split()))
	
	area = [[False]*N for _ in range(M)]
	for _ in range(K):
		x1, y1, x2, y2 = list(map(int,sys.stdin.readline().split()))
		
		for x in range(x1,x2): # 직사각형 범위를 true로 바꿔주어 영역 구할 때 방문하지 못하게
			for y in range(M-y2,M-y1): # y값은 y = M으로 대칭시켜줌(왼쪽 위의 좌표를 (0,0)으로 만들어주기 위함)
				area[y][x] = True
	
	dir = [[-1,0], [1,0], [0,-1], [0,1]] # 상하좌우
	
	a_cnt = [] # 각 영역의 넓이를 담을 배열
	for i in range(M):
		for j in range(N):
			if not area[i][j]: # 직사각형의 범위가 아니거나 방문한 적 없을 때,
				q = collections.deque()
				q.append([i,j])
				area[i][j] = True # bfs 진행을 위해 초기화
				
				cnt = 0 # 현재 영역의 넓이
				while q:
					y, x = q.popleft()
					cnt += 1
					
					for dy, dx in dir:
						dx += x
						dy += y
						
						if dx >= 0 and dx < N and dy >= 0 and dy < M and not area[dy][dx]: # 범위 내에 존재 & 직사각형의 범위 아니고 방문한적도 없음
							q.append([dy,dx])
							area[dy][dx] = True
				
				a_cnt.append(cnt) # 현재 영역의 넓이 push
				
	a_cnt.sort() # 각 영역을 담은 배열 오름차순으로 정렬
	print(len(a_cnt)) # 영역의 수 = 영역을 담은 배열의 길이
	for i in a_cnt: # 정렬된 각 영역의 넓이 출력
		print(i, end = " ")
